sage_code/strong_uniform_bounds.py

```diff
@@ -57,19 +56,27 @@ def is_non_increasing(t, part):
     return True
 
 
-def get_beta_mats_with_pow(F, pow=1):
+def normalize_matrix(m):
+    if m.is_scalar():
+        m2 = m.submatrix(0, 0, 1, 1)
+        m2.set_immutable()
+        return m2
 
-    frob_polys = get_weil_polys(F)
-    output = [matrix.companion(f) ** pow for f in frob_polys]
+    m2 = matrix.companion(m.charpoly()).change_ring(ZZ)
+    m2.set_immutable()
+    return m2
 
-    # We also need to add \pm 1, \pm q
 
-    f1 = (x - 1) * (x + 1)
-    output.append(matrix.companion(f1) ** pow)
+def get_beta_mats_with_pow(F, pow=1, minimize=True):
 
+    frob_polys = get_weil_polys(F)
+    # We also need to add \pm 1, \pm q
     q = F.cardinality()
-    fq = (x - q) * (x + q)
-    output.append(matrix.companion(fq) ** pow)
+    frob_polys += [(x - 1), (x + 1), (x - q), (x + q)]
```

Comment:
these have a different degree to the other frob polys ... you know more about this companion matrix approach to computing the B integers, are we sure that this is OK?

Reply:
Yes, the prime divisors of the bound do not change if we change frob_polys in such a way that the union of the roots of frob polys doesn't change.

I think one important property of the tensor product of matrices is the following: If A and B are matrices that one of the eigenvalues of A is a and one of the eigenvalues of B is b then ab is an eigenvalue of $A \otimes B$. And if $\beta$ is a root of a polynomial $f$ then $\beta$ is an eigenvalue of  matrix.companion(f) (in fact f is the characteristic polynomial of matrix.companion(f). I hope these facts will help you with understanding the entire story. Especially since the determinant of a matrix is just the product of all its eigenvalues counted with multiplicities. This allows you to relate determinants of this tensor products  of companion matrices to norms of products of roots.